server/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java

```diff
@@ -505,6 +551,62 @@ public synchronized void save(Path configDir, char[] password) throws Exception
         }
     }
 
+    /**
+     * Reads the keystore password from the {@link Terminal}, prompting for verification where applicable and returns it as a
+     * {@code char[]}. The caller is responsible to clear the char array after use.
+     *
+     * @param terminal         the terminal to use for user inputs
+     * @param withVerification whether the user should be prompted for password verification
+     * @return a char array with the password the user entered
+     * @throws UserException If the user is prompted for verification and enters a different password
+     */
+    static char[] readPassword(Terminal terminal, boolean withVerification) throws UserException {
+        final char[] password;
+        if (withVerification) {
+            password = terminal.readSecret("Enter new password for the elasticsearch keystore (empty for no password): ");
+            char[] passwordVerification = terminal.readSecret("Enter same password again: ");
+            if (Arrays.equals(password, passwordVerification) == false) {
+                throw new UserException(ExitCodes.DATA_ERROR, "Passwords are not equal, exiting.");
+            }
+            Arrays.fill(passwordVerification, '\u0000');
+        } else {
+            password = terminal.readSecret("Enter password for the elasticsearch keystore : ");
+        }
+        return password;
+    }
+
+    /**
+     * Reads the keystore from disk or attempts to create it if it doesn't already exist. If the keystore is password protected
+     * it also reads the password from user input and decrypts the keystore. Returns both in a POJO. The caller is responsible to clear
+     * the char array with the password after use, calling {@link KeystoreAndPassword#close()}
+     *
+     * @param terminal    the terminal to use for user inputs
+     * @param configFile  the Path from where to attempt and read the existing keystore
+     * @param forceCreate if set, the keystore is created without prompting the user
+     * @return a POJO with the {@link KeyStoreWrapper} and its password, null if the user elected to not create a keystore
+     */
+    static KeystoreAndPassword readOrCreate(Terminal terminal, Path configFile, boolean forceCreate) throws Exception {
```

Comment:
Rather than doing this inside KeyStoreWrapper, I wonder if it could be done in a base keystore cli class (this file is already quite large, and doesn't need to be concerned about cli stuff). Specifically, I think this would eliminate the need for KeystoreAndPassword, and also the need to return null, as to signal to exit. The main method in each CLI would be moved to an abstract method, and this common load/password stuff would be done in the abstract class, putting the keystore and password into member vars of the cli instance, and delegating the rest of the execution to another abstract method. Cleaning up the password and handling a bad password could also be done from this top level, wrapped around the delegate method.

Reply:
Ι went with the KeystoreAndPassword approach after your suggestion in https://github.com/elastic/elasticsearch/pull/37472#discussion_r252450028, but the new suggestion also makes sense and I think I also like it better.